### ISSUE TITLE
(Water) Polygon not filled

### ISSUE THREAD

**Issue by Klaus-Tockloth** [mapsforge/vtm#224]
This water polygon isn't filled in vtm as it does in mapsforge:

Mapsforge:
![bildschirmfoto 2016-11-01 um 16 51 15](https://cloud.githubusercontent.com/assets/5493128/19896317/2a03e560-a054-11e6-9eab-b2744646543d.jpg)

VTM (the eastern part of Lake Zurich is (probably) not filled):
![bildschirmfoto 2016-11-01 um 16 51 31](https://cloud.githubusercontent.com/assets/5493128/19896322/3509ab98-a054-11e6-80bc-15273c5b007e.jpg)

The issue can be reproduces:
Map: http://download.freizeitkarte-osm.de/android/latest/freizeitkarte_che+.map.zip
Theme: https://dl.dropboxusercontent.com/u/1677057/freizeitkarte-vtm.zip

**Comment by devemux86** [comment]
I can confirm that also with Mapsforge [switzerland.map](http://download.mapsforge.org/maps/europe/switzerland.map).

**Comment by Klaus-Tockloth** [comment]
Here's a second example of a missing area:

![bildschirmfoto 2016-11-04 um 20 33 17](https://cloud.githubusercontent.com/assets/5493128/20019762/2369d67a-a2ce-11e6-814c-8b9a871e421a.jpg)

![bildschirmfoto 2016-11-04 um 20 33 27](https://cloud.githubusercontent.com/assets/5493128/20019773/2da72bf6-a2ce-11e6-8fc8-df3020906a3e.jpg)

**Comment by devemux86** [comment]
Probably because of polygons using the default stencil buffer rendering without tessellation, require elements coming from one layer to not overlap.
Using the mesh rendering is slightly slower to load (requires tessellation) and requires more memory, but should be faster to render. And it does not support currently pattern textures.
More details in #212.

So adding `mesh="true"` to area water rule, the lake appears complete.
@Klaus-Tockloth can you test in your theme?

![mesh](https://cloud.githubusercontent.com/assets/3484020/20343120/d7bc2d6e-abf5-11e6-87d8-8daee30e7851.png)

**Comment by Klaus-Tockloth** [comment]
> So adding mesh="true" to area water rule, the lake appears complete.

Yes, this solves the issue:
`<area fill="#80cccc" mesh="true" stroke="#3CCBCB" stroke-width="0.1" />
`

Before:
![bildschirmfoto 2016-11-16 um 18 18 14](https://cloud.githubusercontent.com/assets/5493128/20357791/a31611ae-ac29-11e6-88df-cf25741bddeb.jpg)

After:
![bildschirmfoto 2016-11-16 um 18 19 18](https://cloud.githubusercontent.com/assets/5493128/20357800/ad0b0b06-ac29-11e6-95eb-acf912505ad2.jpg)

**Comment by Klaus-Tockloth** [comment]
The second example is a bit more complex:

![bildschirmfoto 2016-11-16 um 18 16 58](https://cloud.githubusercontent.com/assets/5493128/20357881/fe63132c-ac29-11e6-83d3-e82f549e0b88.jpg)

This is an area (power=sub_station) which is completely surrounded by a fence (barrier=fence). Both objects are included in the map data. The theme renders power=substation as area and barrier=fence as line. Both objects are closed ways. It seems that the fence (drawn as thin line) completely subtracts the sub_station area.

This is the current (incorrect) result:
![bildschirmfoto 2016-11-16 um 18 15 03](https://cloud.githubusercontent.com/assets/5493128/20357997/7408563c-ac2a-11e6-8de4-b9932da73dd0.jpg)

And that would be the expected result:
![bildschirmfoto 2016-11-16 um 18 31 49](https://cloud.githubusercontent.com/assets/5493128/20358151/ffe78204-ac2a-11e6-92e7-13264b20ddaf.jpg)

**Comment by devemux86** [comment]
There was a similar situation in Osmarender with overlapping `natural=scrub` get subtracted with other `natural` areas.
I first added `mesh=true` in b6bfbc612a046aa99e7328ac5e786ce27c581c87 to fix it.

But our other themes do not have that problem, so it's clearly a rules organization issue.
So I separated them in 7c9c3732fb94577cd951717432720176f3263c74 and now the overlapping `natural` areas render without mesh (tessellation).

**Comment by Klaus-Tockloth** [comment]
mesh="true" doesn't work in the (second) case where the area is complete within another area.

**Comment by devemux86** [comment]
@Klaus-Tockloth how is that render theme rule for the 2nd case?

**Comment by Klaus-Tockloth** [comment]
First the area:

```
    <m e="way" k="power" v="plant|generator|substation|station|sub_station" closed="yes">
      <area src="file:/patterns/kraftwerk.svg" symbol-scaling="size" symbol-width="32" />
      <m e="way"   zoom-min="17">
        <caption k="name" style="normal" size="11" fill="#191919" stroke="#00123456" stroke-width="1.0" />
      </m>
    </m>
```

And later the line:

```
    <m e="way" k="barrier" v="fence|wall|city_wall|retaining_wall" zoom-min="16">
      <line stroke="#888888" stroke-width="0.1" />
    </m>
```

**Comment by devemux86** [comment]
Similar(?) case mention in [OpenAndroMaps forum](https://www.openandromaps.org/en/oam-forums/topic/danube-is-half-missing-from-hungary-map).

**Comment by devemux86** [comment]
I enabled for now tessellation on water areas, let's see how it goes.

**Comment by devemux86** [comment]
Lake Constance (Bodensee) with maps of Austria+Switzerland open together:

![constance](https://user-images.githubusercontent.com/3484020/39932926-28fd7878-554a-11e8-9a46-6d5aae3775b7.png)

**Comment by devemux86** [comment]
Performance and stability are more important, as well as proper rendering.

Netherlands before:
![1](https://user-images.githubusercontent.com/3484020/70615406-0448eb00-1c15-11ea-81b0-c9bed8c01d60.png)

Netherlands after:
![2](https://user-images.githubusercontent.com/3484020/70615419-0a3ecc00-1c15-11ea-83d6-65bc26ec1b5d.png)

**Comment by devemux86** [comment]
Fixed in #903.
